### PR TITLE
Table results caching

### PR DESCRIPTION
**Pull request by theopolis**
1. Table implementations (spec files) can mark the table as 'cachable'.
2. Cached results depend on the shortest/quickest interval of scheduled queries that act on results of the table.
3. The table API generator blocks caching on index/additional/required table column options.

### CODE REVIEW

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) failed :-1:

**Comment by marpaia** [comment]
omg! awesome!

File: osquery/core/tables.cpp [hunk 1]
```diff
@@ -136,6 +138,33 @@ PluginResponse TablePlugin::routeInfo() const {
...
+  QueryData results;
+  deserializeQueryDataJSON(content, results);
+  return results;
+}
+
+void TablePlugin::setCache(size_t step,
+                           size_t interval,
+                           const QueryData& results) {
+  // Serialize QueryData and save to database.
+  std::string content;
```
Comment (theopolis) [inline osquery/core/tables.cpp]: This needs a check for cache enablement, otherwise it will be serializing+saving unused data.

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) failed :-1:

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

**Comment by theopolis** [comment]
I was hoping the second commit would reduce time spent in the virtual table filter API, but there's no noticeable performance change/improvement. It does reduce 12k of memory from the binary and from RSS (as all the string data for the ColumnTypes pairs in TablePlugin were constructed).

**Comment by theopolis** [comment]
Not performing string comparisons is just good for business.

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

File: include/osquery/tables.h [hunk 2]
```diff
@@ -309,38 +310,130 @@ typedef struct Constraint Constraint;
...
+   * virtual table APIs. In the best case this context include a limit or
+   * constraints organized by each possible column.
+   *
+   * @param request A query context filled in by SQLite's virtual table API.
+   * @return The result rows for this table, given the query context.
+   */
+  virtual QueryData generate(QueryContext& request) { return QueryData(); }
 
  protected:
+  /// An SQL stable containing the table definition/syntax.
```
Comment (sharvilshah) [inline include/osquery/tables.h]: nit: perhaps meant SQL table instead of SQL stable?

File: include/osquery/tables.h [hunk 3]
```diff
@@ -309,38 +310,130 @@ typedef struct Constraint Constraint;
...
+   * @brief Check if there are fresh cache results for this table.
+   *
+   * Table results are considered fresh when evaluated against a given interval.
+   * The interval is the expected rate for which this data should be generated.
+   * Caching and cache freshness only applies to queries acting on tables
+   * within a schedule. If two queries "one" and "two" both inspect the
+   * table "processes" at the interval 60. The first executed will cache results
+   * and the second will use the cached results.
+   *
+   * Table results are not cached is a QueryContext contains constraints or
```
Comment (sharvilshah) [inline include/osquery/tables.h]: nit: if a QueryContext contains...

**Comment by sharvilshah** [comment]
This is pretty awesome!

**Comment by theopolis** [comment]
Thanks for the review! I'm running the cache on a few machines now to see if it makes much of a difference. I'm expecting A LOT of speed up from the packs we have released, but trying to quantify before releasing this as a default-on option.

**Comment by sharvilshah** [comment]
Not much of a review, just like to keep up with developments and hoping that a second pair of eyes can occasionally be helpful.
And cool, I am also planning on spinning up a couple of EC2 instances this weekend and giving this a spin!

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

**Comment by javuto** [comment]
This is amazing! After this is running, do you think we should drop a bit the frequency in the packs or just leave them as they are?

**Pull request merged**